### PR TITLE
[TeamQuest] Define TeamQuestLevelStatus methods required for progress ca...lculation, define unicode

### PR DESCRIPTION
**Pull request by razvan3895**
@iulianR

### CODE REVIEW

File: wouso/games/teamquest/models.py [hunk 1]
```diff
@@ -132,6 +132,42 @@ def create(cls, status, level):
...
+    @property
+    def points_per_question(self):
+        """ Calculates the rewarded points for a question on a level """
+        total = self.quest_status.levels.all().count() * 1.0 / self.questions.all().count() * 100
+        return total / self.questions.all().count()
+
+    @property
+    def unlocked_questions(self):
+        """ Returns the list of unlocked questions from a level """
+        return TeamQuestQuestion.objects.filter(level=self, locked='U')
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Should be "lock" now

File: wouso/games/teamquest/models.py [hunk 2]
```diff
@@ -95,6 +95,15 @@ def __init__(self, *args, **kwargs):
         # TODO
         super(TeamQuestGame, self).__init__(*args, **kwargs)
 
+    @classmethod
+    def get_current(cls):
+
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Replace this empty line with a docstring. It's not clear what it does without looking at the code

File: wouso/games/teamquest/tests.py [hunk 3]
```diff
@@ -200,6 +208,17 @@ def test_quest_status_create(self):
                 self.assertEqual(team_quest_question.level, level_status)
                 self.assertTrue(team_quest_question.question in level_status.level.questions.all())
 
+    def test_quest_status_progress(self):
+        status = TeamQuestStatus.create(group=self.group, quest=self.quest)
+        self.assertTrue(status.progress == 0)
```
Comment (iulianR) [inline wouso/games/teamquest/tests.py]: Better use assertEqual()

File: wouso/games/teamquest/tests.py [hunk 4]
```diff
@@ -157,15 +157,23 @@ def setUp(self):
...
+        number_of_levels = 5
+        self.questions = []
+        for index in range(number_of_levels * (number_of_levels + 1) / 2):
+            question = Question.objects.create(text='question'+str(index+1), answer_type='F',
+                                               category=category, active=True)
+            self.questions.append(question)
+            answer = Answer.objects.create(text='answer'+str(index+1), correct=True, question=question)
+
+        self.levels = []
+        base = 0
```
Comment (iulianR) [inline wouso/games/teamquest/tests.py]: Add a description for this variable

File: wouso/games/teamquest/tests.py [hunk 5]
```diff
@@ -200,6 +208,17 @@ def test_quest_status_create(self):
                 self.assertEqual(team_quest_question.level, level_status)
                 self.assertTrue(team_quest_question.question in level_status.level.questions.all())
 
+    def test_quest_status_progress(self):
```
Comment (iulianR) [inline wouso/games/teamquest/tests.py]: This only checks the progress when all questions are answered. You should rename this to test_quest_status_progress_100 and write one more test for partial progress.

File: wouso/games/teamquest/models.py [hunk 6]
```diff
@@ -146,6 +204,26 @@ def create(cls, group, quest):
             new_status.levels.add(TeamQuestLevelStatus.create(status=new_status, level=level))
         return new_status
 
+    @property
+    def total_points(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 7]
```diff
@@ -132,6 +154,42 @@ def create(cls, status, level):
...
+        total = self.quest_status.levels.all().count() * 1.0 / self.questions.all().count() * 100
+        return total / self.questions.all().count()
+
+    @property
+    def unlocked_questions(self):
+        """ Returns the list of unlocked questions from a level """
+        return TeamQuestQuestion.objects.filter(level=self, lock='U')
+
+    @property
+    def completed(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 8]
```diff
@@ -132,6 +154,42 @@ def create(cls, status, level):
...
+        return None
+
+    @property
+    def points_per_question(self):
+        """ Calculates the rewarded points for a question on a level """
+        total = self.quest_status.levels.all().count() * 1.0 / self.questions.all().count() * 100
+        return total / self.questions.all().count()
+
+    @property
+    def unlocked_questions(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 9]
```diff
@@ -132,6 +154,42 @@ def create(cls, status, level):
...
+    @property
+    def next_level(self):
+        """ Returns the next_level of a level """
+        for level in self.quest_status.levels.all():
+            if level.index == self.index + 1:
+                return level
+        return None
+
+    @property
+    def points_per_question(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 10]
```diff
@@ -132,6 +154,42 @@ def create(cls, status, level):
...
         return new_level_status
 
+    @property
+    def index(self):
+        """ Unique index of a level in a quest """
+        total_levels = self.quest_status.levels.all().count()
+        return total_levels - self.questions.all().count()
+
+    @property
+    def next_level(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 11]
```diff
@@ -95,6 +95,15 @@ def __init__(self, *args, **kwargs):
         # TODO
         super(TeamQuestGame, self).__init__(*args, **kwargs)
 
+    @classmethod
+    def get_current(cls):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 12]
```diff
@@ -132,6 +154,42 @@ def create(cls, status, level):
             TeamQuestQuestion.create(level=new_level_status, question=question, lock=lock)
         return new_level_status
 
+    @property
+    def index(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

File: wouso/games/teamquest/models.py [hunk 13]
```diff
@@ -115,6 +124,19 @@ def create(cls, level, question, lock):
         new_question = cls.objects.create(level=level, question=question, lock=lock)
         return new_question
 
+    @property
+    def index(self):
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Write a test for it.

**Comment by iulianR** [comment]
Done reviewing! Good job so far.

**Comment by razvan3895** [comment]
@iulianR @razvand 
Wrote tests for everything you said, except for the points_per_question property from TeamQuestLevelStatus. I also changed the progress calculation, making it depend exclusively on the number of questions of a level and the points_per_question property of a level status. My reason for this is that the progress formula could (actually, should) be changed in the future. After my update, if the progress formula changes, only the points_per_question property should be changed.

File: wouso/games/teamquest/tests.py [hunk 14]
```diff
@@ -200,6 +209,152 @@ def test_quest_status_create(self):
...
+        self.levels = []
+        # The start index of the questions sequence that goes in a level
+        base = 0
+        for index in range(number_of_levels):
+            level = TeamQuestLevel.create(quest=None, bonus=0,
+                                          questions=self.questions[base:base+number_of_levels-index])
+            self.levels.append(level)
+            base += number_of_levels - index
+
+        self.quest = TeamQuest.create(title="_test_quest", start_time=datetime.datetime.now(),
```
Comment (iulianR) [inline wouso/games/teamquest/tests.py]: You can just use an empty list of levels considering the test written in this class doesn't care about those.

File: wouso/games/teamquest/tests.py [hunk 15]
```diff
@@ -200,6 +209,152 @@ def test_quest_status_create(self):
...
+
+    def test_get_current_game(self):
+        quest = TeamQuestGame.get_current()
+        self.assertEqual(quest, self.quest)
+
+        self.quest.end_time = datetime.datetime(2010,12,25)
+        self.quest.save()
+
+        quest = TeamQuestGame.get_current()
+        self.assertEqual(quest, None)
```
Comment (iulianR) [inline wouso/games/teamquest/tests.py]: Add a newline

File: wouso/games/teamquest/tests.py [hunk 16]
```diff
@@ -200,6 +209,152 @@ def test_quest_status_create(self):
...
+
+        self.assertEqual(status.progress, status.quest.total_points)
+
+    def test_level_status_index(self):
+        status = TeamQuestStatus.create(group=self.group, quest=self.quest)
+        total_levels = status.levels.all().count()
+        level_indexes = []
+
+        for level_status in status.levels.all():
+            level_questions = level.questions.all().count()
```
Comment (iulianR) [inline wouso/games/teamquest/tests.py]: level should be level_status

**Comment by iulianR** [comment]
Done reviewing.

**Comment by razvan3895** [comment]
@iulianR 
Fixed.

**Comment by iulianR** [comment]
All good!

**Pull request merged**